File: DairoNPC-Android/app/src/main/java/cn/dairo/npc/page/setting/SettingViewModel.kt
package cn.dairo.npc.page.setting

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import cn.dairo.npc.extension.toast
import cn.dairo.npc.repository.NpcRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class SettingViewModel(private val application: Application) : AndroidViewModel(application) {

    private val repo = NpcRepository(application)

    private val _state = MutableStateFlow(
        SettingState()
    )
    val state = this._state.asStateFlow()
    init {
        load()
    }

    /**
     * 暴露更新函数
     */
    fun update(block: (SettingState) -> SettingState){
        this._state.update(block)
    }

    private fun load() {
        viewModelScope.launch {
            _state.update {
                it.copy(
                    npcSetting = repo.loadSetting()
                )
            }
        }
    }
    fun save(block:()-> Unit) {
        var npcSetting = this.state.value.npcSetting
        if(npcSetting.tcpPort.isBlank()){
            npcSetting = npcSetting.copy(tcpPort = "1881")
        }
        if(npcSetting.udpPort.isBlank()){
            npcSetting = npcSetting.copy(udpPort = "1882")
        }

        if(npcSetting.host.isBlank()){
            application.toast("服务器地址不能为空")
            return
        }
        if(npcSetting.key.isBlank()){
            application.toast("秘钥不能为空")
            return
        }
        if(npcSetting.tcpPort.toShortOrNull() == null || npcSetting.tcpPort.toShort() <= 0) {
            application.toast("TCP端口必须为正整数且小于65536")
            return
        }
        if(npcSetting.udpPort.toShortOrNull() == null || npcSetting.udpPort.toShort() <= 0) {
            application.toast("UDP端口必须为正整数且小于65536")
            return
        }
        viewModelScope.launch {
            repo.saveSetting(npcSetting)
            repo.saveSet(true)
            repo.setOpened(true)
            block()
        }
    }
}

File: DairoNPC-IOS/DairoNPC/Page/Setting/SettingViewModel.swift
import Combine
import Foundation
class SettingViewModel: ObservableObject {
    @Published var state = SettingState()
    func save(){
        var npcSetting = self.state.npcSetting
        if(npcSetting.tcpPort.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty){
            npcSetting.tcpPort = "1881"
        }
        if(npcSetting.udpPort.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty){
            npcSetting.udpPort = "1881"
        }
        if(npcSetting.host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty){
            self.state.error = "服务器地址不能为空"
            return
        }
        if(npcSetting.key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty){
            self.state.error = "秘钥不能为空"
            return
        }
        if Int16(npcSetting.tcpPort).map({ $0 > 0 }) != true {
            self.state.error = "TCP端口必须为正整数且小于65536"
            return
        }
        if Int16(npcSetting.udpPort).map({ $0 > 0 }) != true {
            self.state.error = "UDP端口必须为正整数且小于65536"
            return
        }
        NpcRepository.saveSetting(npcSetting)
        NpcRepository.saveSet(true)
        NpcRepository.setOpened(true)
        DairoNPCApp.relaunch()
    }
}
